### ISSUE TITLE
Delay offline warning?

### ISSUE THREAD

**Issue by ojeytonwilliams** [freeCodeCamp/freeCodeCamp#39250]
Would it be better to only show users the offline warning if they've been offline for a few seconds?  Since we retry failed submissions, they should be able to work successfully even if their internet is patchy.

**Comment by raisedadead** [comment]
Yep, could be a good idea.

**Comment by ruhci28** [comment]
hye. can i work on this .?

**Comment by ojeytonwilliams** [comment]
@ruhci28, absolutely!  Anything with the `help wanted` label is available for anyone who wants it.  We typically merge the first accepted PR.

**Comment by ruhci28** [comment]
ok Thanks

**Comment by ruhci28** [comment]
Hey , @ojeytonwilliams  I use setTimeOut function is i am going right.
`import React from 'react';
import PropTypes from 'prop-types';

import './offline-warning.css';

const propTypes = {
  isOnline: PropTypes.bool.isRequired,
  isSignedIn: PropTypes.bool.isRequired
};
const delayInMilliseconds = 5000;
function OfflineWarning({ isOnline, isSignedIn }) {
  return !isSignedIn || isOnline ? null : (
    setTimeout(function() {
      <div className='offline-warning'>
        We cannot reach the server to update your progress.
      </div>
    }, delayInMilliseconds);

);
}

OfflineWarning.displayName = 'OfflineWarning';
OfflineWarning.propTypes = propTypes;

export default OfflineWarning;`

**Comment by ojeytonwilliams** [comment]
Hi @ruhci28 I think the logic should be as follows

```
if offline then set warning timeout
if online then remove warning timeout
```
and
```
if warning timeout resolves then show warning
```

the idea isn't just to delay the warning, it's to only show the warning if the user stays offline for a few seconds.

Does that help at all?

**Comment by ruhci28** [comment]
Thanks @ojeytonwilliams

**Comment by ruhci28** [comment]
Hye , @ojeytonwilliams  does my code makes sense 
import React from 'react';
import PropTypes from 'prop-types';

import './offline-warning.css';

var delayInMilliseconds= 5000;
const propTypes = {
  isOnline: PropTypes.bool.isRequired,
  isSignedIn: PropTypes.bool.isRequired
};
function OfflineWarning({ isOnline, isSignedIn }) {
  const [show_warning,setshow_warning]=React.useState(!isSignedIn || isOnline ? false : true);
  return (
            { (show_warning) && setTimeout(function() {
                 <div className='offline-warning'>
                    You appear to be offline, your progress may not be being saved.
                 </div>
                 },delayInMilliseconds
               )
          }
        )
}

OfflineWarning.displayName = 'OfflineWarning';
OfflineWarning.propTypes = propTypes;

export default OfflineWarning;

**Comment by ojeytonwilliams** [comment]
@ruhci28 are you developing this locally?  The reason I ask is this code has a few syntax errors that your editor should be warning you about.

As for the code itself, I like the direction you're going in.  A couple of points: `setTimeout` returns an `id`, not the result of the callback function.  What you need to do is trigger that timeout if `!isSignedIn || isOnline ? false : true` is `true` and clear the timeout (using that `id`)if it is `false`.  When the timeout resolves, you can use the callback to call `setshow_warning`.  The warning div should then be shown if `show_warning` is `true`.

Finally, using `useState` is a great idea here, but it should be initialised to `false` and the timeout should trigger state changes.

**Comment by ruhci28** [comment]
Thanks @ojeytonwilliams  for the suggestion . yes the editor was giving error and that's why i ask you to review the code . I will implement the changes you tell .

**Comment by ruhci28** [comment]
Hello, @ojeytonwilliams  i made the changes in code can you review it. 
![Screenshot 2020-07-19 at 12 45 20 PM](https://user-images.githubusercontent.com/59759735/87869467-c0457200-c9bd-11ea-8f72-214ddf57fd25.png)

**Comment by ojeytonwilliams** [comment]
@ruhci28 that's looking really good!  The only thing I can see is that as well as clearing the timeout, the state need to be changed.  So, lines 14 to 16 should become:

```
if (!isSignedIn || isOnline) {
  clearTimeout(id);
  if(showWarning) setShowWarning(false);
}
```

This makes sure that the offline warning gets removed once the user is online again.

If you'd like to submit a PR with that, I think we should be able to merge this soon.

**Comment by ruhci28** [comment]
okay i will submit a PR soon
